router/proxy/transport.go

```diff
@@ -138,19 +170,30 @@ type dialErr struct {
 	error
 }
 
-type fakeCloseReadCloser struct {
-	io.ReadCloser
+type fakeCloser struct {
+	io.Closer
 }
 
-func (w *fakeCloseReadCloser) Close() error {
+func (w *fakeCloser) Close() error {
 	return nil
 }
 
-func (w *fakeCloseReadCloser) RealClose() error {
-	if w.ReadCloser == nil {
-		return nil
+func (w *fakeCloser) RealClose() error {
+	return w.Closer.Close()
+}
+
+type runOnFirstRead struct {
+	io.Reader
+
+	fn func() // Run before first Read, then set to nil
+}
+
+func (c *runOnFirstRead) Read(bs []byte) (int, error) {
+	if c.fn != nil {
+		c.fn()
```

Comment:
yeah, I think it cleans this block up substantially to move the `go ...` into `runOnFirstRead.Read()`. Will change it :v: